caffe2/sgd/adagrad_op.cc

```diff
@@ -37,6 +37,30 @@ Optionally returns effective_lr and update as well.
         "Default 1. If it is in (0, 1), the gradient square sum "
         "is decayed by this factor.");
 
+static OpSchema::Cost CostInferenceForSparseAdagrad(
+    const OperatorDef& /* unused */,
+    const vector<TensorShape>& inputs) {
+  CAFFE_ENFORCE_GE(
+      inputs.size(), 4, "SparseAdagrad requires at least 4 inputs");
+
+  const TensorShape param = inputs[0];
+  const TensorShape moment = inputs[1];
+  const TensorShape indices = inputs[2];
+  const TensorShape grad = inputs[3];
+
+  uint64_t n = nElemFromDim(indices);
+  uint64_t grad_size = nElemFromDim(grad);
+
+  OpSchema::Cost c;
+  c.flops = grad_size * 7; // optimistically count sqrt as one flop
```

Comment:
hmm... why is `7`?

Reply:
float hi = nh[i] = h[i] + gi * gi;
    nw[i] = w[i] + lr * gi / (std::sqrt(hi) + epsilon);

Here, we're doing 2 multiplications, 3 additions, 1 division, and 1 sqrt.

Reply:
I see, could you please add this to the comment?